apps/browser/src/popup/services/popup-utils.service.ts

```diff
@@ -115,4 +95,174 @@ export class PopupUtilsService {
+      "popup/index.html#/view-cipher" +
+      "?uilocation=popout" +
+      `&cipherId=${cipherId}` +
+      `&senderTabId=${senderTabId}` +
+      `&action=${action}`;
+
+    await this.openPopout(promptWindowPath, {
+      singleActionPopoutKey: BrowserPopoutType.PasswordReprompt,
+      senderWindowId,
+    });
+  }
+
+  async openAddEditCipherPopout(cipherId: number, senderWindowId: number): Promise<void> {
+    const addEditCipherUrl =
+      cipherId == null
+        ? "popup/index.html#/edit-cipher"
+        : `popup/index.html#/edit-cipher?cipherId=${cipherId}`;
+
+    await this.openPopout(addEditCipherUrl, {
+      singleActionPopoutKey: BrowserPopoutType.AddEditCipher,
+      senderWindowId,
+    });
+  }
+
+  async closeAddEditCipherPopout(delayClose = 0) {
+    await this.closeSingleActionPopout(BrowserPopoutType.AddEditCipher, delayClose);
+  }
+
+  async openTwoFactorAuthPopout(message: { data: string; remember: string }) {
+    const { data, remember } = message;
+    const params =
+      `webAuthnResponse=${encodeURIComponent(data)};` + `remember=${encodeURIComponent(remember)}`;
+    const twoFactorUrl = `popup/index.html#/2fa;${params}`;
+
+    await this.openPopout(twoFactorUrl, { singleActionPopoutKey: BrowserPopoutType.TwoFactorAuth });
+  }
+
+  async closeTwoFactorAuthPopout() {
+    await this.closeSingleActionPopout(BrowserPopoutType.TwoFactorAuth);
+  }
+
+  async openAuthResultPopout(message: { code: string; state: string }) {
+    const { code, state } = message;
+    const authResultUrl = `popup/index.html?uilocation=popout#/sso?code=${encodeURIComponent(
+      code
+    )}&state=${encodeURIComponent(state)}`;
+
+    await this.openPopout(authResultUrl, {
+      singleActionPopoutKey: BrowserPopoutType.SsoAuthResult,
+    });
+  }
+
+  private async openPopout(
+    popupWindowUrl: string,
+    options: {
+      senderWindowId?: number;
+      singleActionPopoutKey?: string;
+      windowOptionsOverrides?: Partial<chrome.windows.CreateData>;
+    } = {}
+  ) {
+    const { senderWindowId, singleActionPopoutKey, windowOptionsOverrides } = options;
+    const defaultPopoutWindowOptions: chrome.windows.CreateData = {
+      type: "normal",
+      focused: true,
+      width: 500,
+      height: 800,
+      ...windowOptionsOverrides,
+    };
+    const offsetRight = 15;
+    const offsetTop = 90;
+    const popupWidth = defaultPopoutWindowOptions.width;
+    const senderWindow = await BrowserApi.getWindow(senderWindowId);
+    const url = chrome.extension.getURL(popupWindowUrl);
+    const parsedUrl = new URL(url);
+    parsedUrl.searchParams.set("uilocation", "popout");
+    if (singleActionPopoutKey) {
+      parsedUrl.searchParams.set("singleActionPopout", singleActionPopoutKey);
+    }
+    const windowOptions = {
+      ...defaultPopoutWindowOptions,
+      url: parsedUrl.toString(),
+      left: senderWindow.left + senderWindow.width - popupWidth - offsetRight,
+      top: senderWindow.top + offsetTop,
+    };
```

Comment:
Non-blocking opinion; if we're allowing window shape to be overridden, we might as well allow position overrides as well. `url` should not be overridden since we want to ensure only URLs in guardrailed scenarios (via `popupWindowUrl`) can be used

```suggestion
    const windowOptions = {
      left: senderWindow.left + senderWindow.width - popupWidth - offsetRight,
      top: senderWindow.top + offsetTop,
      ...defaultPopoutWindowOptions,
      url: parsedUrl.toString(),
    };
```

Reply:
Good call, updated the implementation to fit your suggestion.